### ISSUE TITLE
FR24Feed installation fails when IPv6 is enabled

### ISSUE THREAD

**Issue by fhibler** [jprochazka/adsb-receiver#459]
Hi,

due to an issue with FlightRadar's ability to speak IPv6 the initial configuration of FR24Feed will fail. The issue is described here in detail and forcing `feed.flightradar24.com` to a static IPv4 address does work. The proposed solution tho is far away from ideal.

https://forum.flightradar24.com/threads/11943-Problems-with-feeder-statistics-and-data-sharing?p=106589&viewfull=1#post106589

Best regards,
Florian

**Comment by Romeo-Golf** [comment]
Florian,

While obviously it would be nice if somebody at FR24 can update their DNS to only publish AAAA records when they can actually support it, perhaps you happen to know somebody at CloudFlare who is passionate about IPv6 and could apply some gentle encouragement?

Or failing that remove the AAAA record from feed.flightradar24.com.cdn.cloudflare.net on the basis that broken IPv6 is worse that no IPv6 ;)

Alternatively is the work-around in this post any more acceptable than disabling ipv6, or worse still going back to the pre-dns days of maintain some sort of local hosts.txt file?

https://forum.flightradar24.com/threads/11943-Problems-with-feeder-statistics-and-data-sharing?p=107099&viewfull=1#post107099

**Comment by lucaberta** [comment]
@Romeo-Golf I found myself in the exact same situation after upgrading my Raspi3 from Jessie to Stretch.

It's also worth noticing that I have a fully-functional IPv6 network, still the connection was failing, and a telnet connection to feed.flightradar24.com from the CLI worked well.

The issue was also being shown on the log with a weird failure in the loopback connection from fr24feed to modesmixer2, which I use locally to dispatch different data streams to different feeds, splitting and reformatting the raw data coming from my USB-connected ModeS Beast.

To me, this is clearly a bug and should be troubleshooted in order to understand what causes the issue. I am not a big fan of dirty workarounds.

To fix the issue, I used a trick which you call "back to the pre-DNS days" but it's in fact a common way to fix (or break!) connections if needed, by overriding DNS replies. This trick is used by many viruses to trick antiviruses into failing their updates when "calling home"!

At any rate, this is the message that I just wrote on the FR24 forum:

FYI, I too have run into connection issues, which might be related to IPv6, which is strange as I have a full IPv6-enabled network at home, and the connection would work the same via IPv4 or IPv6. Go figure.

Nonetheless, it was clear that the issue was related to IPv6 when I decided to create a simple hack which would force IPv4-only connections, of a very different kind than blocking completely IPv6 (which I DON'T want to do!) and neither changing the configuration of the getaddrinfo config file /etc/gai.conf.

The hack consists of fixing the IP address of the initial connection to feed.flightradar24.com and its CloudFlare counterpart to a fixed IPv4 address, which will take precedence over a DNS-resolved IP address.

Here is what I added in /etc/hosts:

```
# hack to fix the IPv6 issue with feed.flightradar24.com.cdn.cloudflare.net (104.20.0.101) - solved on 2018-08-25

104.20.0.101 feed.flightradar24.com
104.20.0.101 feed.flightradar24.com.cdn.cloudflare.net
```

Obviously I now have a fixed IP address and not a dynamic one out of a pool, and if FR24 someday decides to change the IP address of the feed, I will need to modify it by hand.

By doing this I was able to make my newly refreshed RasPi 3, which used to run Jessie and now runs Stretch, work again after the dist-upgrade which crippled fr24feed.

Bye, Luca

**Comment by fhibler** [comment]
@Romeo-Golf Agreed. I've done the 'hosts' workaround as well, although still you are locking yourself in. Should CF decide to change the IP on the CDN, this would cause an issue (as @lucaberta mentioned). I wonder if there would be some way to just 'force' the FR24Feed connection via v4. Unfortunately I'm not really helpful here, as I don't have any ideas besides '/etc/hosts' or a local DNS which just fails AAAA at all.

Bests,
Florian

**Comment by Romeo-Golf** [comment]
While I havent tested it myself, the following workaround from the FR24 thread seems like its may have the least impact:

> Perhaps a better solution to those who still want to run both IPv4 and IPv6 on their Pi's:
> 
> In the file /etc/gai.conf find the following line:
> 
> #precedence ::ffff:0:0/96 100
> 
> Uncomment this line so it looks like:
> 
> precedence ::ffff:0:0/96 100
> 
> Your Pi will now prefer IPv4 over IPv6 without disabling IPv6.
> Both FR24 feeder and FlightAware feeder should now be happy and you don't have to worry about any DNS changes/updates.
> 
> Hope this helps!
> 
> Edit: If you have previously disabled IPv6 then you need to re-enable it for this to work properly!
> 

Alternatively if FR havent either started to accept connections on IPv6, or had Cloudflare remove the AAAA records feed.flightradar24.com.cdn.cloudflare.net, could it be worth asking Cloudflare to ask their "customer" if they really want the AAAA record which is breaking their non-IPv6 enabled application? ;)

**Comment by lucaberta** [comment]
@Romeo-Golf it depends whether you consider switching back to a default of IPv4 as “least impact” or not.

As I mentioned in my message, I did not want to touch the getaddrinfo preference because I would rather use IPv6 than v4, and I did not consider that option as a workaround because one shortcoming of FR24 would impact the behavior of my whole machine.

Workarounds are dirty, we know that, but they should have the least amount of impact possible, and IMHO using a hardwired address in /etc/hosts is by far the simplest way to solve this annoying issue on our end.

Ideally, as a short term fix Cloudflare could remove the AAAA records, until FR24 fixes things with their IPv6 support, as you say.

**Comment by risticsasa011** [comment]
I, actually, prefer the solution with the /etc/gai.conf file change, as it can be a permanent solution. 

I don't care whether ipv4 or ipv6 is used, as long as the app works, and the ipv6 address is reachable from the global internet. This is, almost, a valid fix, since changing the /etc/hosts file will, sooner or later, result in a broken feed....

**Comment by lucaberta** [comment]
@risticsasa011 the only issue with changing the /etc/gai.conf file is that it will impact **every** process running on the machine and accessing the network. And I don't want that to happen.

I much prefer a "sooner or later" on a single non-critical daemon like fr24feed rather than having a much bigger impact on the whole machine.

**Comment by risticsasa011** [comment]
@lucaberta 
well, it will impact every process, indeed... but it will work, period :) 
the other way around, we have a broken feed 100%... 
I don't see why would this be a big issue, the rest of the processes should only try ipv4 first, and if there is no ipv4 connectivity, they should failover to ipv6... 

I am aware that it is a step back in a very wrong direction... I just don't see this having any big impact, could you elaborate?

**Comment by lucaberta** [comment]
@risticsasa011 no, if one uses /etc/hosts everything works great, as long as the IP address of the FR24 server does not change. And I don’t know how often it changes, really. On top of things, one could write a simple script to query DNS using a type A query and updating the /etc/hosts every once in a while. But it’s really overkill.

Why don’t Ilike to change /etc/gai.conf? Because I try to use IPv6 as much as I can, since I have a full IPv6 enabled network and my ISP fully supports IPv6.

At any rate, workarounds are what they are, workarounds. They are always ugly, and sometimes lead to other types of failure.

Best thing would be if fr24feed would be fixed!

**Comment by Romeo-Golf** [comment]
+1, im not registered on the fr24 forum but if you are I would encourage you to complain loudly and frequently that since their move to CloudFlare they have broken their data feed submission for a non-trivial percentage of their userbase!
